x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/ElasticServiceAccounts.java

```diff
@@ -49,6 +49,12 @@ final class ElasticServiceAccounts {
                         ".elastic-connectors*"
                     )
                     .privileges("manage", "read", "write")
+                    .build(),
+                // Connector secrets. Enterprise Search can read and write to this index.
+                RoleDescriptor.IndicesPrivileges.builder()
```

Comment:
I'm confused about the privileges management. What we used in the past is a bit different:
- Define new privileges (like `WRITE_CONNECTOR_SECRETS` and `READ_CONNECTORS_SECRETS`). These privileges are used to restrict actions for users based on the defined patterns (like [here](https://github.com/elastic/elasticsearch/blob/770671667eaad7eca8404785c0422bb54a4213f5/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java#L157)).
- Assign these privileges to the service accounts (or the users that need them), like [here](https://github.com/elastic/elasticsearch/blob/0fca9c3098d446b865ab5071f9b0f29982c32a5f/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/ElasticServiceAccounts.java#L65).
- Create internal users for managing the system indices. It's there that the index-level privileges over the system index are defined, like [here](https://github.com/elastic/elasticsearch/blob/c390d666628c425a6355a3e3fc58cfc0fa785d3c/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java#L164).

Does that make sense? Or is this a conscious decision and I'm missing the point?

Reply:
>  There was no point adding one for read/write for Enterprise Search because it has manage as a cluster privilege which gives it full access to the cluster already.

AFAIU `manage` does not provide actual index read permission for the index, even if it provides the tools for eventually having that.

> There shouldn't be a situation (at least in 8.13) where a non-Enterprise Search user needs read access to secrets. That's why I've not included a cluster privilege for it.

If we have created a cluster privilege for Kibana write, why not create a cluster privilege for reading? That seems odd to me.

> When are the internal users "used" in practice?

They are converted to system privileges. Connectors use the internal XPack user [here](https://github.com/elastic/elasticsearch/blob/c390d666628c425a6355a3e3fc58cfc0fa785d3c/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java#L155), which [has full access to indices](https://github.com/elastic/elasticsearch/blob/770671667eaad7eca8404785c0422bb54a4213f5/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUsers.java#L98) IIUC.

This means that using `CONNECTORS_ORIGIN` for a client internally translates it to the `XPACK_USER`, which has full access to system indices including security.

The problem with the current approach is that any user that has access to the get secret action will execute the action using the `OriginSettingClient`, which internally translates to the XPack user, who has access to the system index. As there is no privilege for reading secrets, any user that access the endpoint will end up being able to read secrets. That's a security issue.

The way to change that would be to either:
- Define a read privilege that would limit access to the get secrets action, or
- Do not use the `OriginSettingClient`, but use the actual user that makes the request and rely on their index permissions for security.

I find defining privileges is better and more explicit than relying on system indices permissions.

I hope this makes sense. If I have misunderstood the security mechanism please call me out - but I would do a quick check with a non-privileged user just to make sure that accessing the secrets is not possible.

Reply:
Chiming in on this thread, I know that for fleet secrets we have a `write_fleet_secrets` method and that's the pattern we reuse here. If we look at the other built in cluster privileges, the pattern we usually have is to have a `manage_[feature]` privilege paired with a `read_[feature]` privilege:

```
GET _security/privilege/_builtin

response:
{
  "cluster": [
    "all",
    "cancel_task",
    "create_snapshot",
    "cross_cluster_replication",
    "cross_cluster_search",
    "delegate_pki",
    "grant_api_key",
    "manage",
    "manage_api_key",
    "manage_autoscaling",
    "manage_behavioral_analytics",
    "manage_ccr",
    "manage_data_frame_transforms",
    "manage_enrich",
    "manage_ilm",
    "manage_index_templates",
    "manage_ingest_pipelines",
    "manage_logstash_pipelines",
    "manage_ml",
    "manage_oidc",
    "manage_own_api_key",
    "manage_pipeline",
    "manage_rollup",
    "manage_saml",
    "manage_search_application",
    "manage_search_query_rules",
    "manage_search_synonyms",
    "manage_security",
    "manage_service_account",
    "manage_slm",
    "manage_token",
    "manage_transform",
    "manage_user_profile",
    "manage_watcher",
    "monitor",
    "monitor_data_frame_transforms",
    "monitor_enrich",
    "monitor_ml",
    "monitor_rollup",
    "monitor_snapshot",
    "monitor_text_structure",
    "monitor_transform",
    "monitor_watcher",
    "none",
    "post_behavioral_analytics_event",
    "read_ccr",
    "read_fleet_secrets",
    "read_ilm",
    "read_pipeline",
    "read_security",
    "read_slm",
    "transport_client",
    "write_fleet_secrets"
  ],
  "index": [
    "all",
    "auto_configure",
    "create",
    "create_doc",
    "create_index",
    "cross_cluster_replication",
    "cross_cluster_replication_internal",
    "delete",
    "delete_index",
    "index",
    "maintenance",
    "manage",
    "manage_data_stream_lifecycle",
    "manage_follow_index",
    "manage_ilm",
    "manage_leader_index",
    "monitor",
    "none",
    "read",
    "read_cross_cluster",
    "view_index_metadata",
    "write"
  ]
}
```

Maybe we should do the same here - have `manage_connectors_secrets` and `read_connectors_secrets`.
`manage_connectors_secrets` allows for both reads and writes. 
I think this is a better pattern and I am thinking how App Search also has a read/write pattern for API keys. In retrospect I'd wish we went for the same pattern with manage/read.

Reply:
> The problem with the current approach is that any user that has access to the get secret action will execute the action using the OriginSettingClient, which internally translates to the XPack user, who has access to the system index. As there is no privilege for reading secrets, any user that access the endpoint will end up being able to read secrets. That's a security issue.

We should have a test for this use case.

Reply:
@carlosdelest @ioanatia I've added the `read_connector_secrets` cluster privilege back in [this commit](https://github.com/elastic/elasticsearch/pull/103683/commits/9cf3f81a9e115dd26c15e24124d087a944ea4543). I've opted not to use `manage_` because this is a little different from our other cluster privileges. I want to have `write_` as a separate privilege because some roles should have write without read permissions.

If we were to use `manage_`, the two cluster privileges would be `write_connector_secrets` and `manage_connector_secrets`. I'm okay with that but I wasn't sure if it aligned with what yous were expecting. I'm happy to change it to `manage_` and `write_`, or keep `read_` and `write_`. WDYT?

> The problem with the current approach is that any user that has access to the get secret action will execute the action using the OriginSettingClient, which internally translates to the XPack user, who has access to the system index. As there is no privilege for reading secrets, any user that access the endpoint will end up being able to read secrets. That's a security issue.
>> We should have a test for this use case.

Do you know of any example tests I can use for reference? I'm not sure what I should even be looking for here.

Reply:
AH sorry, I see you've linked examples in [this comment](https://github.com/elastic/elasticsearch/pull/103683#discussion_r1447101822). I'll try to add some tests for this.

Reply:
okay, I tested this locally with your branch, it seems we do require `manage` or `all` which I think is on purpose since we don't want to risk adding APIs that don't require a privilege:

```
# with no roles
 curl -XGET "http://localhost:9200/_connector/_secret/-R7z94wBGenV8ZXMozwE" -H "kbn-xsrf: reporting" -u ioana:changeme
{"error":{"root_cause":[{"type":"security_exception","reason":"action [cluster:admin/xpack/connector/secret/get] is unauthorized for user [ioana] with effective roles [], this action is granted by the cluster privileges [manage,all]"}],"type":"security_exception","reason":"action [cluster:admin/xpack/connector/secret/get] is unauthorized for user [ioana] with effective roles [], this action is granted by the cluster privileges [manage,all]"},"status":403}%
# with an unrelated role
curl -XGET "http://localhost:9200/_connector/_secret/-R7z94wBGenV8ZXMozwE" -H "kbn-xsrf: reporting" -u ioana:changeme
{"error":{"root_cause":[{"type":"security_exception","reason":"action [cluster:admin/xpack/connector/secret/get] is unauthorized for user [ioana] with effective roles [machine_learning_user], this action is granted by the cluster privileges [manage,all]"}],"type":"security_exception","reason":"action [cluster:admin/xpack/connector/secret/get] is unauthorized for user [ioana] with effective roles [machine_learning_user], this action is granted by the cluster privileges [manage,all]"},"status":403}
```

It would still be good to add an explicit test that a user without `manage` or `all` cannot access the API, just to catch any regressions.

>  > manage_connectors_secrets allows for both reads and writes.

> I want the user/Kibana to only have write access however. If we use manage_ is it okay to keep it restricted to write only?

Then we would just change the name, which doesn't do much IMO. I don't have enough context to understand why we would want to restrict the user/Kibana to only writes.
For the purpose of this PR I would say:
- let's use `manage` and `all` for writes/deletes
- add tests to check that a user without any of these privileges cannot do writes/deletes on connectors secrets.

Then in a separate PR we can see what we can do about an additional privilege like `write_connectors_secret` /`read_connectors_secret`. WDYT?

Reply:
@ioanatia @carlosdelest I've added tests and fixed up some of the cluster privilege stuff in [this commit](https://github.com/elastic/elasticsearch/pull/103683/commits/54875c354496fdb9a0d2519a32c08ad7e3f2283e).

@ioanatia I'm not sure how much of your comment still applies as we discussed this PR on slack since then. Do you still feel strongly about using `manage` and `all` for writes/deletes? Or have we now covered the issues you mentioned?

Reply:
Thanks for iterating on this @navarone-feekery ! 💯